### ISSUE TITLE
Re-think operation generators

### ISSUE THREAD

**Issue by josh146** [PennyLaneAI/pennylane#970]
Currently in PennyLane, qubit operations define their generator in the following form:

```python
operation.generator = [G, coeff]
```

where `G` can be either

* a PennyLane operation *class*, e.g., `qml.PauliX`,  or
* a numeric NumPy matrix,

and `coeff` is the coefficient in the matrix exponential `exp(1j * coeff * G * param)`.

The generators are only used for metric tensor computations, and this approach served us well back when we only had single-qubit rotations. However, our gate set has grown to include multi-qubit parametrized gates that can be applied to an *arbitrary* number of wires. This has caused two issues:

1. How do we define the generator for a multi-parameter gate such as `qml.Rot`?
2. How do we best define the generator for a multi-wire gate such as `qml.MultiRZ`, `qml.PauliRot`, `qml.CRot`, etc?

(1) is probably easy to support, we just need to think about the matrix exponential form of these gates, and generalize the generator property to support linear combinations.

(2) we currently get around by defining the generator as a NumPy matrix (see #963), but this is not ideal and scales pretty badly. Instead, rather than returning an operation class or a matrix, we should instead allow returning an operation *instance* or a matrix.

That way we could support returning operators defined over the operation wires, which avoids creating a large matrices. For example, for `qml.PauliRot("XY", wires=[0, 2])`, we should support:

```python
operation.generator = [qml.PauliX(wires=self.wires[0]) @ qml.PauliY(wires=self.wires[1]), coeff]
```

**Comment by dwierichs** [comment]
I think it would be valuable to decide whether generator should

1. allow for multiple entries in the list, where each operator is unitary and the sum forms the actual generator _or_
2. allow for non-unitary entries.

This would be relevant for gates like `CRZ`, which is not generated by a unitary.

**Comment by josh146** [comment]
> This would be relevant for gates like CRZ, which is not generated by a unitary.

Very good point @dwierichs.

I'm very keen to fix how generators are implemented, since quite a lot of functionality under-the-hood depends on it. Currently, this means that these features are restricted to working with only the RX, RY, and RZ gates.

**Comment by dwierichs** [comment]
I think there are plenty of possibilities to do this.. nice that you want to expand on this!
When thinking of less elementary gates, multiparameter gates might not even take the form e^{-i\theta P} e^{-i\phi Q} ,right?
But I guess it is reasonable to extend the generator-based features to those gates for which the notion of _the_ generator for each parameter is well-defined, and worry about even more general cases later.
It might be noteworthy that just to cover the first option of the two I suggested above and multi-parameter gates at the same time, `generator` already needs to have a structure like
```python
generator = ( 
    [ [ first_par_first_op, first_par_first_op_coeff ], [ first_par_sec_op, first_par_sec_op_coeff ] ],
    [ [ sec_par_first_op, sec_par_first_op_coeff ], [ sec_par_sec_op, sec_par_sec_op_coeff ] ],
    )
```

**Comment by co9olguy** [comment]
more intelligent generators are definitely a must-have, e.g., for the adjoint method implementation on arbitrary gates :+1:

**Comment by josh146** [comment]
Yep; in particular, for the adjoint method, we need the derivative of the operation with respect to the parameter we are differentiating:

![image](https://user-images.githubusercontent.com/2959003/106995726-75a11100-67ba-11eb-958c-2d94f00a205e.png)

Currently we do this in PL with a _lot_ of assumptions:

```python
generator, prefactor = operation.generator
return 1j * prefactor * generator.matrix @ operation.matrix
```

The new framework for generators we come up with should easily scale and allow the operation derivative to be computed for more operations beyond RX, RY, RZ.

**Comment by dwierichs** [comment]
As a (potentially infeasible) fall-back solution for single-parameter gates, one could compute the derivative at 0 to get the generator.

Also, for large gates, allowing for sparse data type generators would be good.

**Comment by josh146** [comment]
> As a (potentially infeasible) fall-back solution for single-parameter gates, one could compute the derivative at 0 to get the generator.

I thought about this as well 😆 But I think this should definitely be a last resort, I can imagine it adding a lot of overhead during optimization as opposed to having the generators statically available

**Comment by josh146** [comment]
An update on this issue: @mariaschuld rightly points out that a concept of a generator only makes sense for gates of the form `U(t) = e^{i G t}` (e.g., one-parameter gates).

For the adjoint and fisher information, we don't need the generator so much as we need the *partial derivative of the operation*, for each parameter, evaluated at parameter=0.

So perhaps we should not conflate these, and instead:

* Improve the current generator API, but restrict it to single-parameter operations
* Provide an API for easily retrieving the value `∂_θᵢ U(θ) |_{θᵢ=0}`

**Comment by dwierichs** [comment]
>@mariaschuld rightly points out that a concept of a generator only makes sense for gates of the form U(t) = e^{i G t} (e.g., one-parameter gates).
Yeah, that's of course true!
>For the adjoint and fisher information, we don't need the generator so much as we need the partial derivative of the operation, for each parameter, evaluated at 0.
Why is the derivative only ever needed at 0? If a multi-parameter gate does not consist of successively applied single-parameter gates, but for example `U(s,t)=e^{i (sG + tH)}`, this might be more complicated, right?

**Comment by josh146** [comment]
I believe this can now be closed with #2030 merged!
